### ISSUE TITLE
Celery worker using 100% CPU around epoll w/ prefork+redis and not consuming tasks

### ISSUE THREAD

**Issue by bpowers** [celery/celery#1845]
Hello,

We're using Celery 3.1.6, and one of our queues is using 100% CPU for the master process.  ps shows:

    root      6960 59.2  0.7 160680 51628 ?        Rs   Feb01 2319:49 /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO
    root     32727  0.3  1.1 266008 82860 ?        S    Feb02   6:29  \_ /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO
    root     32729  0.2  1.1 265184 82040 ?        S    Feb02   5:54  \_ /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO
    root     32730  0.2  1.4 282868 99788 ?        S    Feb02   5:29  \_ /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO
    root       750  0.1  1.5 296584 113100 ?       S    Feb02   3:03  \_ /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO
    root       917  0.2  1.1 262044 78888 ?        S    Feb02   6:00  \_ /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO
    root      1873  0.0  1.2 269460 86276 ?        S    Feb02   1:34  \_ /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO
    root      1875  0.0  1.4 285568 102384 ?       S    Feb02   0:36  \_ /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO
    root      2347  0.0  0.8 245448 60984 ?        S    Feb02   1:47  \_ /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO
    root      2348  0.0  0.8 247760 63356 ?        S    Feb02   1:33  \_ /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO
    root      2362  0.0  1.0 257880 73416 ?        S    Feb02   1:09  \_ /$VENV/bin/python /$VENV/manage.py celery worker -Q ads -P prefork --autoscale=10,5 -n $NAME --settings=settings -l INFO

And strace is more interesting.  It shows the following loop as the only thing happening in the parent:

    gettimeofday({1391450996, 406286}, NULL) = 0
    epoll_wait(4, {{EPOLLIN, {u32=6, u64=22205092589469702}}}, 1023, 1000) = 1
    epoll_ctl(4, EPOLL_CTL_DEL, 34, {EPOLLRDNORM|EPOLLRDBAND|EPOLLWRNORM|EPOLLMSG|0x4e9020, {u32=0, u64=22205092589469696}}) = -1 ENOENT (No such file or directory)
    epoll_ctl(4, EPOLL_CTL_DEL, 67, {EPOLLRDNORM|EPOLLRDBAND|EPOLLWRNORM|EPOLLMSG|0x4e9020, {u32=0, u64=22205092589469696}}) = -1 ENOENT (No such file or directory)
    epoll_ctl(4, EPOLL_CTL_DEL, 68, {EPOLLRDNORM|EPOLLRDBAND|EPOLLWRNORM|EPOLLMSG|0x4e9020, {u32=0, u64=22205092589469696}}) = -1 ENOENT (No such file or directory)
    epoll_ctl(4, EPOLL_CTL_DEL, 38, {EPOLLRDNORM|EPOLLRDBAND|EPOLLWRNORM|EPOLLMSG|0x4e9020, {u32=0, u64=22205092589469696}}) = -1 ENOENT (No such file or directory)
    epoll_ctl(4, EPOLL_CTL_DEL, 10, {EPOLLRDNORM|EPOLLRDBAND|EPOLLWRNORM|EPOLLMSG|0x4e9020, {u32=0, u64=22205092589469696}}) = -1 ENOENT (No such file or directory)
    epoll_ctl(4, EPOLL_CTL_DEL, 53, {EPOLLRDNORM|EPOLLRDBAND|EPOLLWRNORM|EPOLLMSG|0x4e9020, {u32=0, u64=22205092589469696}}) = -1 ENOENT (No such file or directory)
    epoll_ctl(4, EPOLL_CTL_DEL, 26, {EPOLLRDNORM|EPOLLRDBAND|EPOLLWRNORM|EPOLLMSG|0x4e9020, {u32=0, u64=22205092589469696}}) = -1 ENOENT (No such file or directory)
    epoll_ctl(4, EPOLL_CTL_DEL, 59, {EPOLLRDNORM|EPOLLRDBAND|EPOLLWRNORM|EPOLLMSG|0x4e9020, {u32=0, u64=22205092589469696}}) = -1 ENOENT (No such file or directory)
    epoll_ctl(4, EPOLL_CTL_DEL, 30, {EPOLLRDNORM|EPOLLRDBAND|EPOLLWRNORM|EPOLLMSG|0x4e9020, {u32=0, u64=22205092589469696}}) = -1 ENOENT (No such file or directory)
    epoll_ctl(4, EPOLL_CTL_DEL, 63, {EPOLLRDNORM|EPOLLRDBAND|EPOLLWRNORM|EPOLLMSG|0x4e9020, {u32=0, u64=22205092589469696}}) = -1 ENOENT (No such file or directory)
    epoll_ctl(4, EPOLL_CTL_ADD, 51, {EPOLLIN|EPOLLERR|EPOLLHUP, {u32=51, u64=22205092589469747}}) = -1 EEXIST (File exists)
    epoll_ctl(4, EPOLL_CTL_ADD, 52, {EPOLLIN|EPOLLERR|EPOLLHUP, {u32=52, u64=22205092589469748}}) = -1 EEXIST (File exists)
    epoll_ctl(4, EPOLL_CTL_ADD, 6, {EPOLLIN|EPOLLERR|EPOLLHUP, {u32=6, u64=22205092589469702}}) = -1 EEXIST (File exists)

Those FDs are pipes according to `/proc/$PID/fd`.  (strace also shows heartbeat/worker chatter from other processes read via the redis socket).

Restarting the upstart job caused the worker to begin consuming tasks again.  I think this is some corner case that isn't being handled well.  Kombu's eventio.py has the following:

```python
class _epoll(Poller):
    # ...
    def register(self, fd, events):
        try:
            self._epoll.register(fd, events)
        except Exception as exc:
            if get_errno(exc) != errno.EEXIST:
                raise
    def unregister(self, fd):
        try:
            self._epoll.unregister(fd)
        except (socket.error, ValueError, KeyError):
            pass
        except (IOError, OSError) as exc:
            if get_errno(exc) != errno.ENOENT:
                raise
```

We are specifically checking for ENOENT and EEXIST here and silently dropping the exception - it seems like this might be causing a problem for someone further down the stack, but I'm not familiar with this part of celery.

Any help/pointers appreciated.  Let me know what other info is useful as well.

**Comment by gfranxman** [comment]
@swuensch That appears to be part of a larger object.    How is that used?

**Comment by swuensch** [comment]
@gfranxman The method shown is part of a singleton object that tracks all currently running tasks in its self.running_tasks field. The signal handler calls the shown revoke_all_tasks() method of this object.

Anyways, important for the purpose of the workaround is just the call of Celery's Task.revoke() method with these parameters, "relaxing" its asynchronous behavior:
```
 t.revoke(terminate=True, wait=True, timeout=1)
```

**Comment by DevJac** [comment]
I was experiencing something similar to this bug. I was seeing 100% CPU usage on the master worker process while most of the worker subprocesses sat idle. In plain words, it was like the tasks were getting stuck.

I tried using the `-Ofair` flag when starting the worker and this appears to be an effective workaround in my case. I hadn't seen this mentioned specifically in this thread, so I'm mentioning it now. I'd consider this is a workaround rather than a fix.

A little more detail:
I started my worker with a command like `celery worker -A proj -Ofair`. See the `celery worker --help` page for more info (although it doesn't specifically document the `-Ofair` flag).

**Comment by joostdevries** [comment]
@ask @bpowers we're seeing this issue as well and right now it happens about once a day on our production stack. I'm considering switching back over to RabbitMQ but since recent reports also demonstrate the issue there I'm not sure that'll help.

Any progress on tracking this issue down? From my logs it appears to happen **after** the consumption of a task (because the prerun is logged but the postrun is never logged). 

If there's anything I can do to help please let me know

**Comment by leetreveil** [comment]
@joostdevries I work at the same company as @bpowers and we have worked around this issue by setting the redis client timeout to a value greater than our slowest task:

```
redis-cli config set timeout 7200
```

**Comment by joostdevries** [comment]
@leetreveil Our task takes about 1-2 sec so I can't imagine this would be a timeout issue. It's set to 300 right now..

**Comment by shshank** [comment]
Having similar issue.

celery==3.1.17
celery-with-redis==3.0

@joostdevries I dont think this is happening after execution of a task. I restarted the worker and even without task it slowly crept to ~99% CPU.

**Comment by shshank** [comment]
-Ofair didnt help either.

**Comment by ask** [comment]
I notice you are revoking tasks here, but doing so in an extremely ineffective way.

Revoking several tasks at once like this:

    [result.revoke() for result in results]

should be replaced by:

    app.control.revoke([result.id for result in results])


or `GroupResult.revoke()`

**Comment by kakiezhang** [comment]
i have the similar situation. when i use prefork ( 5 Queues and each 8 Concurrency ) pool type to run celery workers in 8 machine, the redis machine's cpu will be up to 100% usage in peak times. 

i shutdown 1 machine, then it's working well, and the redis machine's cpu seems to be normal ( 6% ~ 10%)

i also check the slowlog info in redis, it's find that lpush and brpop 's execution time is up to 12 ~ 20ms. 

the brpop's timeout = 1, so i doubt that if i created that money workers, they occupied resource to get the key. Some get, some timeout, even some of them exceed the timeout, and then they might be wait a long time util they can get the key again.

so, i decide to cut down the process to let the resource is not being so nerves. 

is there anything i can do to help it work?

**Comment by chase-seibert** [comment]
Just started running into this as well. Switching to -P gevent fixes it for me.

**Comment by dcramer** [comment]
We're seeing this on some AWS infrastructure running Sentry:

- Sentry 8.0.0.dev0
- Celery 3.1.18
- Kombu 3.0.26
- Redis 2.10.x

We tried a huge number of things, and nothing seemed to help, but the symptoms are identical to what the original ticket describes (though this issue seems to have gone way off base, as it has nothing to do with revocation or gevent etc). The primary symptom we originally saw was workers stalling for 30 minutes every hour or so. We couldn't identify what the 30 minutes was, but now the best guess is it was due to the Redis timeout value. After several attempts we eventually got workers to stall var more frequently.

Things attempted, and failed miserably at, in a number of combinations:

- Set socket_timeout to 60 on the broker
- Set timeout to 0 on redis (vs 1800)
- Disable mingle, gossip, heartbeat (which, btw, seems to break celery entirely, not sure why this is an option)

Of note, Sentry defaults to disabling many options in Celery including rate limits, results, and we don't use any of the complex features.

Along the way we also increased worker concurrency from something like 8 to 32. This pointed out that Celery doesn't handle load balancing well. Unrelated, but something to note.

We aren't certain it's fully resolved, but here's what we're currently looking at, which at least got the queue caught up and was stable for 1h+:

- Redis timeout of 1800
- socket_timeout not set (whatever default is)
- Worker running with -O fair --without-mingle --without-gossip
- Soft timeout of 60s

After enabling the above in one go, we saw it churning through things as well as saw various timeouts in tasks. These timeouts were to the same Redis node in seemingly at least two situations:

- Communication via Sentry's internal TSDB
- (Unconfirmed due to empty stacktrace) a call to .delay()

My bet is that the resolution for us came from -O fair and lowering the prefect. It's not clear to me how or why, unless we are inducing far too much load on a single Redis instance and that again is causing the symptoms shown in this ticket.

**Comment by bivald** [comment]
I'm seeing the same issue on celery 3.1.18 but without using Redis. Using Google Compute Engine. Hangs at 100% cpu and not consuming more tasks (or in some cases, only 1 of 20 threads)

**Comment by domenkozar** [comment]
Did you print strace for each worker? This is similar to https://github.com/celery/celery/issues/2606 (except that I didn't observe 100% CPU usage)

**Comment by ask** [comment]
@dcramer @domenkozar Do you have the `fanout_prefix` and `fanout_patterns` transport options set? http://docs.celeryproject.org/en/latest/getting-started/brokers/redis.html#caveats

**Comment by dcramer** [comment]
No. Even if we were, the issue of "literally everything stops working" is less so of a caveat than a critical bug.

We use very basic features. A dozen queues, all direct. A few periodic tasks. Anything else we tested with and without and it didnt help. We also disable rate limits, task results, and task events.

**Comment by dcramer** [comment]
@ask if we can get a fix in Kombu for this can we get a bugfix release out that doesn't require the newer Django stuff? Sentry isn't upgrading past Django 1.6.

**Comment by ask** [comment]
You definitely should be setting those options, or the worker will be receiving multiple event messages for every task executed by every worker.

However, the 100% CPU issue is probably not related to that.  I'm guessing that some socket is closed/broken by the remote end, but not removed from the event loop.  When a socket is broken it is always readable, so the event loop spins on it.

It could be there is some place in redis-py/kombu where an exception is swallowed, and the connection automatically reconnected or something, but the event loop is never notified.

**Comment by ask** [comment]
@dcramer: celery 3.1 should be compatible with Django 1.6, if not it would be a bug

**Comment by mitsuhiko** [comment]
@ask I opened up a new ticket #2906 which covers the issue that @dcramer talked about here.  From further investigation it looks like it might not be related to this one as I could never reproduce actual high CPU load.

**Comment by cralston-prism** [comment]
For anyone who stumbles across this thread in the future: Upgrade kombu to 3.0.33 or newer.

**Comment by markddavidoff** [comment]
still experiencing this issue with celery 3.1.25 and kombu 3.0.33...

**Comment by auvipy** [comment]
this issue should be fixed in celery 4.2.1+ version. the version you are using is not supported anymore and don't contain most of the improvements and bug fixes were done in past 2+ years.

**Comment by PavanKulkarni** [comment]
@auvipy Most of the bug fixes mentioned here are present  <4.1.0. any specific reason why you mention >4.2.1?
